src/main/java/io/numaproj/numaflow/function/FunctionService.java

```diff
@@ -284,4 +241,59 @@ private Udfunction.DatumList buildDatumListResponse(MessageT[] messageTs) {
         });
         return datumListBuilder.build();
     }
+
+    /*
+        extracts and returns the result to the observer.
+     */
+    private void extractResult(List<Future<Object>> futureList, StreamObserver<Udfunction.DatumList> responseObserver, ActorRef supervisorActorRef) {
+        Udfunction.DatumList.Builder responseBuilder = Udfunction.DatumList.newBuilder();
+
+        // build the response when its completed.
+        Futures
+                .sequence(futureList, actorSystem.dispatcher())
+                .onComplete(new OnComplete<>() {
+                    @Override
+                    public void onComplete(
+                            Throwable failure,
+                            Iterable<Object> success) {
+
+                        // if there are any failures indicate it to the observer.
+                        if (failure != null) {
+                            log.error("error while getting output from actors - "
+                                    + failure.getMessage());
+
+                            responseObserver.onError(failure);
+                            return;
+                        }
+
+                        success.forEach(ele -> {
+                            Udfunction.DatumList list = buildDatumListResponse((Message[]) ele);
+                            responseBuilder.addAllElements(list.getElementsList());
+                        });
+                        Udfunction.DatumList response = responseBuilder.build();
+
+                        responseObserver.onNext(response);
+                        responseObserver.onCompleted();
+
+                        /*
+                            once the result is return we can stop the supervisor actor.
+                            stopping the supervisor will stop all its child actors.
+                            we should  explicitly stop the actors for it to be garbage collected.
+                        */
+                        actorSystem.stop(supervisorActorRef);
```

Comment:
> stopping the supervisor will stop all its child actors.

Is `shutdownActorRef` also its child actor? If not, do we need to stop it?

Reply:
ah right! we need to stop it..

Reply:
Did we address this one?

Reply:
can we reuse shutdownActorRef ?

Reply:
I handled it inside shut down actor's completedSuccessfully method.

Reply:
I think we can reuse it, it doesn't have any dependency on a particular partition.

Reply:
My bad, we can't reuse it, it has a dependency on the response observer, which is different for each stream.